akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -179,22 +180,38 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
               incomingStreams(streamId).inlet.foreach(_.cancel())
 
             case WindowUpdateFrame(streamId, increment) ⇒
-              incomingStreams(streamId).outboundWindowLeft += increment
-              debug(f"outbound window for [$streamId%3d] is now ${incomingStreams(streamId).outboundWindowLeft}%10d after increment $increment%6d")
-              bufferedFrameOut.tryFlush()
+              try {
+                incomingStreams(streamId).outboundWindowLeft += increment
+
+                println(s"incomingStreams(streamId) = ${incomingStreams(streamId)} ( added + $increment ) ")
+                debug(f"outbound window for [$streamId%3d] is now ${incomingStreams(streamId).outboundWindowLeft.value}%10d after increment $increment%6d")
+                bufferedFrameOut.tryFlush()
+              } catch {
+                case ex: ArithmeticException ⇒
+                  pushGOAWAY(ErrorCode.FLOW_CONTROL_ERROR, debug = ex.getMessage)
+              }
 
             case PriorityFrame(streamId, exclusiveFlag, streamDependency, weight) ⇒
               debug(s"Received PriorityFrame for stream $streamId with ${if (exclusiveFlag) "exclusive " else "non-exclusive "} dependency on stream $streamDependency and weight $weight")
 
             case SettingsFrame(settings) ⇒
-              debug(s"Got ${settings.length} settings!")
+              if (settings.nonEmpty) debug(s"Got ${settings.length} settings!")
+
               settings.foreach {
                 case Setting(Http2Protocol.SettingIdentifier.SETTINGS_INITIAL_WINDOW_SIZE, value) ⇒
                   debug(s"Setting initial window to $value")
-                  val delta = value - streamLevelWindow
-                  streamLevelWindow = value
-                  incomingStreams.values.foreach(_.outboundWindowLeft += delta)
-                case Setting(id, value) ⇒ debug(s"Ignoring setting $id -> $value")
+                  val delta = WindowSizeWithOverflowProtection(value) - streamLevelWindow.value
+                  streamLevelWindow = WindowSizeWithOverflowProtection(value)
+                  incomingStreams.values.foreach(_.outboundWindowLeft += delta.value)
+
+                case Setting(Http2Protocol.SettingIdentifier.SETTINGS_MAX_FRAME_SIZE, value) ⇒
+                  debug(s"Already set client max frame size to ${value} in framing stage...")
+
+                case tableSizeSetting @ Setting(Http2Protocol.SettingIdentifier.SETTINGS_HEADER_TABLE_SIZE, _) ⇒
+                  debug("Bouncing header table size update back to encoder...")
+                  bufferedFrameOut.push(SyntheticHpackEncoderSettingFrame(tableSizeSetting))
+                case Setting(id, value) ⇒
+                  debug(s"Ignoring setting $id -> $value")
               }
 
               bufferedFrameOut.push(SettingsAckFrame)
```

Comment:
Add the received settings here to the ack frame instead of the specialized bouncing frames.